<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" id="sixapart-standard">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="generator" content="Movable Type Pro 5.13-en" />
<link rel="stylesheet" href="/pub/styles.css" type="text/css" />
<link rel="start" href="/pub/" title="Home" />
<link rel="alternate" type="application/atom+xml" title="Recent Entries" href="/pub/atom.xml" />
<script type="text/javascript" src="/pub/mt.js"></script>
<!--
<rdf:RDF xmlns="http://web.resource.org/cc/"
         xmlns:dc="http://purl.org/dc/elements/1.1/"
         xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
<Work rdf:about="/pub/2001/11/16/perlbio2.html">
<dc:title>Parsing Protein Domains with Perl</dc:title>
<dc:description> The Perl programming language is popular with biologists because of its practicality. In my book, Beginning Perl for Bioinformatics, I demonstrate how many of the things biologists want to write programs for are readily-even enjoyably-accomplished with Perl. My book...</dc:description>
<dc:creator>James D. Tisdall</dc:creator>
<dc:date>2001-11-16T00:00:00-08:00</dc:date>
<license rdf:resource="http://creativecommons.org/licenses/by-nc-nd/3.0/" />
</Work>
<License rdf:about="http://creativecommons.org/licenses/by-nc-nd/3.0/">
</License>
</rdf:RDF>
-->

<script type="text/javascript">

 var _gaq = _gaq || [];
 _gaq.push(['_setAccount', 'UA-50555-22']);
 _gaq.push(['_trackPageview']);

 (function() {
   var ga = document.createElement('script'); ga.type =
   'text/javascript'; ga.async = true;
   ga.src = ('https:' == document.location.protocol ? 'https://ssl' :
   'http://www') + '.google-analytics.com/ga.js';
   var s = document.getElementsByTagName('script')[0];
   s.parentNode.insertBefore(ga, s);
 })();

</script>
<script type='text/javascript' src='http://partner.googleadservices.com/gampad/google_service.js'></script>
<script type='text/javascript'>
GS_googleAddAdSenseService("ca-pub-4136420132070439");
GS_googleEnableAllServices();
</script>
<script type='text/javascript'>
GA_googleAddSlot("ca-pub-4136420132070439", "Perl_728x90");
</script>
<script type='text/javascript'>GA_googleFetchAds();</script>
    
    <link rel="prev bookmark" href="/pub/2001/11/15/creatingrss.html" title="Create RSS channels from HTML news sites" />
    <link rel="next bookmark" href="/pub/2001/11/21/lightweight.html" title="Lightweight Languages" />
    
    
    <title>Parsing Protein Domains with Perl - Perl.com</title>
</head>
<body id="perl-com" class="mt-entry-archive layout-wt">
    <div id="container">
        <div id="container-inner">


            <div id="header">
    <div id="header-inner">
        <div id="header-content">
        <span id="top_advert"> 
<!-- Put any landscape advert in here -->
<!-- Perl_728x90 -->
<script type='text/javascript'>
GA_googleFillSlot("Perl_728x90");
</script>
        </span> 



            <div id="header-name"><a href="/pub/" accesskey="1">Perl.com</a></div>
            <div id="header-description">news and views of the Perl programming language</div>




        </div>
    </div>
</div>



            <div id="content">
                <div id="content-inner">


                    <div id="alpha">
                        <div id="alpha-inner">


                            <div id="entry-1630" class="entry-asset asset hentry">
                                <div class="asset-header">
                                    <h1 id="page-title" class="asset-name entry-title">Parsing Protein Domains with Perl</h1>
                                    <div class="asset-meta">
                                        <span class="byline">

                                            By <span class="vcard author">James D. Tisdall</span> on <abbr class="published" title="2001-11-16T00:00:00-08:00">November 16, 2001 12:00 AM</abbr>

                                        </span>


                                    </div>
                                </div>
                                <div class="asset-content entry-content">

                                    <div class="asset-body">
                                        
<!-- sidebar begins -->
<!-- don't move sidebars -->
<!-- sidebar ends -->


<p>
The Perl programming language is popular with biologists because of its practicality. In my book, 
<a href="http://www.oreilly.com/catalog/begperlbio/">Beginning Perl for 
Bioinformatics</a>, I demonstrate how many of the things 
biologists want to write programs for are readily--even enjoyably--accomplished with Perl.</p>

<p>My book teaches biologists how to program in Perl, even
if they have never programmed before. This article will use Perl at the level found in the middle-to-late chapters in my book, after some of the basics have been learned. However, this article can be read by biologists who do not (yet) know any programming. They should be able
to skim the program code in this article, only reading the comments, to get a general feel for how Perl is used in practical applications, using real biological data. </p>

<p>
Biological data on computers tends to be either in structured ASCII flat files--that is to say, in plain-text files--or in relational databases. Both of these data sources are easy to handle with Perl programs. For this article, I will discuss one of the flat-file data sources, 
the <a href="http://ca.expasy.org/prosite/" target="new">Prosite database</a>, which contains valuable biological information about protein domains. I will demonstrate how to use Perl to extract and use the protein domain information. In <i>Beginning Perl for 
Bioinformatics</i> I also show how to work with several other similar data sources, including GenBank (Genetic Data Bank), PDB (Protein
DataBank), BLAST (Basic Local Alignment Search Tool)
output files, and REBASE (Restriction Enzyme Database).</p>

<h3>What is Prosite?</h3>



<p>
Prosite stands for "A Dictionary of Protein Sites and Patterns." To learn more about the fascinating biology behind Prosite, visit the 
<a href="http://ca.expasy.org/cgi-bin/lists?prosuser.txt">Prosite User Manual</a>. Here's an introductory description of Prosite from the user manual:</p>

<blockquote>
 <p>"Prosite is a method of determining what is the function of uncharacterized proteins translated from genomic or cDNA sequences. It consists of a database of biologically significant sites and patterns formulated in such a way  that with  appropriate  computational  tools  it  can  rapidly  and reliably identify  to which known family of protein (if any) the new sequence belongs."</p>

<p>In some cases, the sequence of an unknown protein is too distantly related to any protein of known structure to detect its resemblance by overall sequence alignment. However, it can be identified by the occurrence in its sequence of a particular cluster of residue types, variously known as a pattern, a motif, a signature, or a fingerprint. These motifs arise because of particular requirements on the structure of specific regions of a protein, which may be important, for example, for their binding properties, or for their enzymatic activity.</p>
</blockquote>



Prosite is available as a set of plain-text files that provide the data, plus documentation. The <a href="http://www.expasy.ch/prosite">Prosite home 
page</a> provides a user interface that allows you to query the database and 
examine the documentation. The database can also be obtained for local installation from the 
<a href="ftp://www.expasy.ch/databases/prosite">Prosite ftp 
site</a>. Its use is free of charge for noncommercial users. </p>

<p>
There is some fascinating and important biology involved here; and in the programs that
follow there are interesting and useful Perl programming techniques.  See the Prosite
User Manual for the biology background, and <i>Beginning Perl for Bioinformatics</i> for the
programming background.  Or just keep reading to get a taste for what is possible when
you combine programming skills with biological data.</p>

<h3>Prosite Data</h3>

<p>
The Prosite data can be downloaded to your computer. It is in the ASCII flat file called 
<a href="ftp://ca.expasy.org/databases/prosite/release_with_updates/prosite.dat">prosite.dat</a> and is more than 4MB in size. A small version of this file created for this article, called <i>prosmall.dat</i>, is available 
<a href="http://perl.com/2001/11/16/examples/prosmall.dat">here</a>. This version of the data has just the first few records from the complete file, making it easier for you to download and test, and it's the file that we'll use in the code discussed later in this article. </p>

<p>
Prosite also provides an accompanying data file, 
<a href="ftp://ca.expasy.org/databases/prosite/release_with_updates/prosite.doc">prosite.doc</a>, which contains documentation for all the 
records in <i>prosite.dat</i>. Though we will not use it for this article, I 
do recommend you look at it and think about how to use the information along with the code presented here if you plan on doing more with Prosite.</p>


<blockquote>
<hr size="1" noshade="noshade" />
<blockquote>
<table border="0" cellpadding="0" cellspacing="0" height="70">
<tr><td valign="center" height="70">
<p class="secondary"><a href="http://conferences.oreilly.com/biocon/"><img 
src="http://oreilly.com/graphics_new/biocon_logo.gif" align="right"
hspace="0" vspace="0" border="0" width="120" height="70" 
alt="O'Reilly Bioinformatics Technology Conference" /></a>
James Tisdall will be speaking at O'Reilly's first 
Bioinformatics Technology Conference, January 28-31, 2002, in Tuscon, 
Arizona. For more information visit <a 
href="http://conferences.oreilly.com/biocon/">Bioinformatics Conference 
Web site</a>.  
</p></td></tr></table>
</blockquote>
<hr size="1" noshade="noshade" />
</blockquote>
<p>
The Prosite data in <a 
href="ftp://ca.expasy.org/databases/prosite/release_with_updates/prosite.dat" 
target="new">prosite.dat</a> (or our much smaller test file <i>prosmall.dat</i>) is organized in "records," each 
of which consists of several lines, and which always include an ID line and a 
termination line containing "//". The Prosite lines all begin with a 
two-character code that specifies the kind of data that appears on that line. 
Here's a breakdown of all the possible line types that a record may contain from the 
<a href="http://ca.expasy.org/cgi-bin/lists?prosuser.txt" 
target="new">Prosite User Manual</a>: </p>

<em>
<dl>
<dt>ID</dt><dd>Identification (Begins each entry; one per entry)</dd>
<dt>AC</dt><dd>Accession number (one per entry)</dd>
<dt>DT</dt><dd>Date (one per entry)</dd>
<dt>DE</dt><dd>Short description (one per entry)</dd>
<dt>PA</dt><dd>Pattern (>=0 per entry)</dd>
<dt>MA</dt><dd>Matrix/profile (>=0 per entry)</dd>
<dt>RU</dt><dd>Rule (>=0 per entry)</dd>
<dt>NR</dt><dd>Numerical results (>=0 per entry)</dd>
<dt>CC</dt><dd>Comments (>=0 per entry)</dd>
<dt>DR</dt><dd>Cross references to SWISS-PROT (>=0 per entry)</dd>
<dt>3D</dt><dd>Cross references to PDB (>=0 per entry)</dd>
<dt>DO</dt><dd>Pointer to the documentation file (one per entry)</dd>
<dt>//</dt><dd>Termination line (Ends each entry; one per entry)</dd>
</dl>
</em>

<p>
Each of these line types has certain kinds of information that are formatted in a specific manner, as is detailed in the Prosite documentation.
</p>

<h3>Prosite Patterns</h3>

<p>
Let's look specifically at the Prosite patterns. These are presented in a 
kind of mini-language that describes a set of short stretches of protein that 
may be a region of known biological activity. Here's the description of the 
pattern "language" from the <a 
href="http://ca.expasy.org/cgi-bin/lists?prosuser.txt">Prosite User 
Manual</a>: </p>

<blockquote>
<p>
The PA (PAttern) lines contains the definition of a Prosite pattern. The 
patterns are described using the following conventions: </p>	

<ul>
<li>The standard IUPAC one-letter codes for the amino acids are used.</li>

<li>The symbol `x' is used for a position where any amino acid is 
accepted.</li>

<li>Ambiguities are indicated by listing the acceptable amino acids for a
given position, between square parentheses `[  ]'. For example: [ALT]
stands for Ala or Leu or Thr.</li>

<li>Ambiguities are also indicated by listing between a pair of curly
brackets `{  }' the amino acids that are not accepted at a given position. 
For example: {AM} stands for any amino acid except Ala and Met.</li>

<li>Each element in a pattern is separated from its neighbor by a `-'.</li>

<li>Repetition of an element of the pattern can be indicated by following
that element  with a numerical value or a numerical range between 
parenthesis. Examples: x(3) corresponds to x-x-x, x(2,4) corresponds to
x-x or x-x-x or x-x-x-x.</li>

<li>When a pattern is restricted to either the  N- or C-terminal of a 
sequence, that pattern either starts with a `<' symbol or respectively
ends with a `>' symbol.</li>

<li>A period ends the pattern.</li>
</ul>
</blockquote>

<h4>Perl Subroutine to Translate Prosite Patterns into Perl Regular 
Expressions</h4>

<p>
In order to use this pattern data in our Perl program, we need to translate 
the Prosite patterns into Perl regular expressions, which are the main way 
that you search for patterns in data in Perl. 
For the sake of this article I will assume that you know the basic regular 
expression syntax. (If not, just read the program comments, and skip the Perl 
regular expressions.) As an example of what the following subroutine does, 
it will translate the Prosite pattern <code>[AC]-x-V-x(4)-{ED}.</code> into the 
equivalent Perl regular expression <code>[AC].V.{4}[^ED]</code> </p> 


<p>
Here, then, is our first Perl code, the subroutine <code>PROSITE_2_regexp</code>, to 
translate the Prosite patterns to Perl regular expressions: </p>

<p>
<pre>
<code>
#
# Calculate a Perl regular expression
#  from a PROSITE pattern
#
sub PROSITE_2_regexp {

  #
  # Collect the PROSITE pattern
  #
  my($pattern) = @_;

  #
  # Copy the pattern to a regular expression
  #
  my $regexp = $pattern;

  #
  # Now start translating the pattern to an
  #  equivalent regular expression
  #

  #
  # Remove the period at the end of the pattern
  #
  $regexp =~ s/.$//;

  #
  # Replace 'x' with a dot '.'
  #
  $regexp =~ s/x/./g;

  #
  # Leave an ambiguity such as '[ALT]' as is.
  #   However, there are two patterns [G>] that need
  #   special treatment (and the PROSITE documentation
  #   is a bit vague, perhaps).
  #
  $regexp =~ s/\[G\>\]/(G|\$)/;
  
  #
  # Ambiguities such as {AM} translate to [^AM].
  #
  $regexp =~ s/{([A-Z]+)}/[^$1]/g;

  #
  # Remove the '-' between elements in a pattern
  #
  $regexp =~ s/-//g;

  #
  # Repetitions such as x(3) translate as x{3}
  #
  $regexp =~ s/\((\d+)\)/{$1}/g;

  #
  # Repetitions such as x(2,4) translate as x{2,4}
  #
  $regexp =~ s/\((\d+,\d+)\)/{$1}/g;

  #
  # '<' becomes '^' for "beginning of sequence"
  #
  $regexp =~ s/\&lt;/^/;

  #
  # '>' becomes '$' for "end of sequence"
  #
  $regexp =~ s/\&gt;/\$/;

  #
  # Return the regular expression
  #
  return $regexp;
}
</code>
</pre>
</p>

<p>
Subroutine <code>PROSITE_2_regexp</code> takes the Prosite pattern and translates 
its parts step by step into the equivalent Perl regular expression, as 
explained in the comments for the subroutine. If you do not know Perl regular 
expression syntax at this point, just read the comments--that is, the lines 
that start with the # character. That will give you the general idea of the 
subroutine, even if you don't know any Perl at all. </p>

<blockquote>
<hr size="1" noshade="noshade" />
<blockquote>
<p class="secondary">Learn more about the power of regular expressions from O'Reilly's 
<a href="http://www.oreilly.com/catalog/regex/">Mastering Regular 
Expressions: Powerful Techniques for Perl and Other Tools</a>. </p>
</blockquote>
<hr size="1" noshade="noshade" />
</blockquote>















<h4>Perl Subroutine to Parse Prosite Records into Their Line Types</h4>

<p>
The other task we need to accomplish is to parse the various types of lines, so that, for 
instance, we can get the ID and the PA pattern lines easily. The next 
subroutine accomplishes this task: given a Prosite record, it returns a 
hash with the lines of each type indexed by a key that is the 
two-character "line type". The keys we'll be interested in 
are the ID key for the line that has the identification information; and 
the PA key for the line(s) that have the pattern information. </p>

<p>
This "get_line_types" subroutine does more than we need. It makes a hash 
index on all the line types, not just the ID and PA lines that we'll 
actually use here. But that's OK. The subroutine is short and simple 
enough, and we may want to use it later to do things with some of the other types of lines in a Prosite record.</p>

<p>
By building our hash to store the lines of a record, we can extract any of the data lines from the record that we like, just by giving the line type code (such as ID for identification number). We can use this hash to 
extract two line types that will interest us here, the ID identifier line 
and the PA pattern line. Then, by translating the Prosite pattern into a 
Perl regular expression (using our first subroutine), we will be in a 
position to actually look for all the patterns in a protein sequence. In 
other words, we will have extracted the pattern information and made it 
available for use in our Perl program, so we can search for the patterns in the protein sequence. </p>

<blockquote>
<hr size="1" noshade="noshade" />
<blockquote>
<p class="secondary">If you're interested in learning Perl, don't miss O'Reilly's best-selling 
<a href="http://www.oreilly.com/catalog/lperl3/">Learning Perl, 3rd 
Edition</a>, which has been updated to cover Perl version 5.6 and rewritten 
to reflect the needs of programmers learning Perl today. For a complete 
list of O'Reilly's books on Perl, go to <a 
href="http://perl.oreilly.com">perl.oreilly.com</a>. </p>
</blockquote>
<hr size="1" noshade="noshade" />
</blockquote>
<p>
Here, then, is our second subroutine, which accepts a Prosite record, and 
returns a hash which has the lines of the record indexed by their line 
types: </p>

<pre><code>
#
# Parse a PROSITE record into "line types" hash
# 
sub get_line_types {

  #
  # Collect the PROSITE record
  #
  my($record) = @_;

  #
  # Initialize the hash
  #   key   = line type
  #   value = lines
  #
  my %line_types_hash = ();

  #
  # Split the PROSITE record to an array of lines
  #
  my @records = split(/\n/,$record);

  #
  # Loop through the lines of the PROSITE record
  #
  foreach my $line (@records) {

    #
    # Extract the 2-character name
    # of the line type
    #
    my $line_type = substr($line,0,2);

    #
    # Append the line to the hash
    # indexed by this line type
    #
    (defined $line_types_hash{$line_type})
    ?  ($line_types_hash{$line_type} .= $line)
    :  ($line_types_hash{$line_type} = $line);
  }

  #
  # Return the hash 
  #
  return %line_types_hash;
}
</code></pre>

<h3>Main Program</h3>

<p>
Now let's see the code at work. The following program uses the subroutines we've just
defined to read in the Prosite records one at a time from the database in the flat file
prosmall.txt.  It then separates the different kinds of lines (such as "PA" for
pattern), and translates the patterns into regular expressions, using the subroutine
PROSITE_2_regexp we already wrote.  Finally, it searches for the regular expressions
in the protein sequence, and reports the position of the matched pattern in the sequence.</p>

<pre><code>
#!/usr/bin/perl
#
# Parse patterns from the PROSITE database, and
# search for them in a protein sequence
#

#
# Turn on useful warnings and constraints
#
use strict;
use warnings;

#
# Declare variables
#

#
# The PROSITE database
#
my $prosite_file = 'prosmall.dat';

#
# A "handle" for the opened PROSITE file
#
my $prosite_filehandle; 

#
# Store each PROSITE record that is read in
#
my $record = '';

#
# The protein sequence to search
# (use "join" and "qw" to keep line length short)
#
my $protein = join '', qw(
MNIDDKLEGLFLKCGGIDEMQSSRTMVVMGGVSG
QSTVSGELQDSVLQDRSMPHQEILAADEVLQESE
MRQQDMISHDELMVHEETVKNDEEQMETHERLPQ
);

#
# open the PROSITE database or exit the program
#
open($prosite_filehandle, $prosite_file)
 or die "Cannot open PROSITE file $prosite_file";

#
# set input separator to termination line //
#
$/ = "//\n";

#
# Loop through the PROSITE records
#
while($record = <$prosite_filehandle>) {

  #
  # Parse the PROSITE record into its "line types"
  #
  my %line_types = get_line_types($record);

  #
  # Skip records without an ID (the first record)
  #
  defined $line_types{'ID'} or next;

  #
  # Skip records that are not PATTERN
  # (such as MATRIX or RULE)
  #
  $line_types{'ID'} =~ /PATTERN/ or next;

  #
  # Get the ID of this record
  #
  my $id = $line_types{'ID'};
  $id =~ s/^ID   //;
  $id =~ s/; .*//;

  #
  # Get the PROSITE pattern from the PA line(s)
  #
  my $pattern = $line_types{'PA'};
  # Remove the PA line type tag(s)
  $pattern =~ s/PA   //g;

  #
  # Calculate the Perl regular expression
  # from the PROSITE pattern
  #
  my $regexp =  PROSITE_2_regexp($pattern);

  #
  # Find the PROSITE regular expression patterns
  # in the protein sequence, and report
  #
  while ($protein =~ /$regexp/g) {
    my $position = (pos $protein) - length($&) +1;
    print "Found $id at position $position\n";
    print "   match:   $&\n";
    print "   pattern: $pattern\n";
    print "   regexp:  $regexp\n\n";
  }
}

#
# Exit the program
#
exit;
</code>
</pre>


<p>
This program is available online as the file <a 
href="http://perl.com/2001/11/16/examples/parse_prosite">parse_prosite</a>.
The tiny example Prosite database is available as the file <a 
href="http://perl.com/2001/11/16/examples/prosmall.dat">prosmall.dat</a>. If you save these files on your 
(Unix, Linux, Macintosh, or Windows) computer, you can enter the following 
command at your command-line prompt (in the same folder in which you saved 
the two files): </p>

<pre><code>
% perl parse_prosite
</code></pre>

<p>
and it will produce the following output:
</p>

<pre><code>
Found PKC_PHOSPHO_SITE at position 22
   match:   SSR
   pattern: [ST]-x-[RK].
   regexp:  [ST].[RK]

Found PKC_PHOSPHO_SITE at position 86
   match:   TVK
   pattern: [ST]-x-[RK].
   regexp:  [ST].[RK]

Found CK2_PHOSPHO_SITE at position 76
   match:   SHDE
   pattern: [ST]-x(2)-[DE].
   regexp:  [ST].{2}[DE]

Found MYRISTYL at position 30
   match:   GGVSGQ
   pattern: G-{EDRKHPFYW}-x(2)-[STAGCN]-{P}.
   regexp:  G[^EDRKHPFYW].{2}[STAGCN][^P]
</code>
</pre>

<p>
As you see, our short program goes through the Prosite database one record at a time, parsing each record according to the types of lines within it. If the record has an ID and a pattern, it then extracts them, creates a Perl regular expression from the pattern, and finally searches in a protein sequence for the regular expression, reporting on the patterns found. </p>

<h3>The Next Step</h3>

<p>
This article has shown you how to take biological data from the Prosite 
database and use it in your own programs. With this ability, you can write programs specific to your particular research needs.</p>

<p>
Many kinds of data discovery are possible: you could combine searches for Prosite patterns with some other computation. For instance, you may want to 
also search the associated genomic DNA or cDNA for restriction sites 
surrounding a particular Prosite pattern in the translated protein, in 
preparation for cloning. </p>
<blockquote>
<hr size="1" noshade="noshade" />
<blockquote>
<p class="secondary">James Tisdall has also written
<a href="http://www.oreilly.com/news/perlbio_1001.html">Why Biologists 
Want to Program Computers</a> for <i>oreilly.com</i>.
</blockquote>
<hr size="1" noshade="noshade" /> </p>
</blockquote>
<p>
While such programs are interesting in their own right, their importance in 
laboratory research really lies in the fact that their use can save enormous 
amounts of time; time which can then be used for other, less routine, tasks 
on which biological research critically depends. </p>

<p>
This article gives an example of using Perl to extract and use data from a flat file database, of which there are many in biological research. In fact, some of the most important biological databases are in flat file format, including GenBank and PDB, the primary databases for DNA sequence information and for protein structures.</p>

<p>
With the ability to write your own programs, the true power of 
bioinformatics can be applied in your lab. Learning the Perl programming language can give you a direct entry into this valuable new laboratory technique. </p>

<!--Steve: will you add short bio here? -->

<hr size="1" noshade>

<p class="secondary">
O'Reilly &amp; Associates recently released (October 2001) <a 
href="http://oreilly.com/catalog/begperlbio/">Beginning Perl for 
Bioinformatics</a>.
</p>

<ul>
<li><p class="secondary">
<a href="http://www.oreilly.com/catalog/begperlbio/chapter/ch10.html">Sample 
Chapter 10, GenBank</a>, is available free online. 
</p></li>

<li><p class="secondary">
You can also look at the <a 
href="http://www.oreilly.com/catalog/begperlbio/toc.html">Table of 
Contents</a>, the 
<a href="http://www.oreilly.com/catalog/begperlbio/toc.html">Index</a>, and 
the <a href="http://www.oreilly.com/catalog/begperlbio/desc.html">Full 
Description</a> of the book.
</p></li>

<li><p class="secondary">
For more information, or to order the book, <a 
href="http://oreilly.com/catalog/begperlbio/">click here</a>.</p></li>  
</ul>





















                                    </div>


                                </div>
                                <div class="asset-footer">


                                    <div class="entry-tags">
                                        <h4>Tags<span class="delimiter">:</span></h4>
                                        <ul>
                                            <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=Beginning%20Perl%20for%20Bioinformatics&amp;limit=20';return false;" rel="tag">Beginning Perl for Bioinformatics</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=bioinformaticians&amp;limit=20';return false;" rel="tag">bioinformaticians</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=bioinformaticists&amp;limit=20';return false;" rel="tag">bioinformaticists</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=bioinformatics&amp;limit=20';return false;" rel="tag">bioinformatics</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=Bioinformatics&amp;limit=20';return false;" rel="tag">Bioinformatics</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=biological%20research&amp;limit=20';return false;" rel="tag">biological research</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=biology%20research%20programming&amp;limit=20';return false;" rel="tag">biology research programming</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=Computational%20biology&amp;limit=20';return false;" rel="tag">Computational biology</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=genetics&amp;limit=20';return false;" rel="tag">genetics</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=genomics&amp;limit=20';return false;" rel="tag">genomics</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=James%20Tisdall&amp;limit=20';return false;" rel="tag">James Tisdall</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=learning%20to%20program&amp;limit=20';return false;" rel="tag">learning to program</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=O%27Reilly&amp;limit=20';return false;" rel="tag">O'Reilly</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=o%27reilly&amp;limit=20';return false;" rel="tag">o'reilly</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=oreilly&amp;limit=20';return false;" rel="tag">oreilly</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=Perl%20for%20bioinformatics&amp;limit=20';return false;" rel="tag">Perl for bioinformatics</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=programming%20for%20biology%20research&amp;limit=20';return false;" rel="tag">programming for biology research</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=sequencing&amp;limit=20';return false;" rel="tag">sequencing</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=tisdall&amp;limit=20';return false;" rel="tag">tisdall</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?IncludeBlogs=2&amp;tag=Web&amp;limit=20';return false;" rel="tag">Web</a></li>
                                        </ul>
                                    </div>

                                </div>
                            </div>


                    
                    


                        </div>
                    </div>


                    <div id="beta">
    <div id="beta-inner">


    
    <div class="widget-what-is-perl widget">
    <div class="widget-content widget-content-what-is-perl">
       Visit the home of the  Perl programming language: <a href="http://www.perl.org/">Perl.org</a
    </div>
</div>
<div class="widget-find-out-more widget-archives widget">
    <div class="widget-content">
        <ul>
            <li><a href="http://www.perl.org/get.html">Download</a></li>
            <li><a href="http://perldoc.perl.org/">Documentation</a></li>
            <li><a href="http://blogs.perl.org/">Perl Bloggers</a></li>
            <li><a href="http://news.perlfoundation.org/">Foundation News</a></li>
        </ul>
    </div>
</div><div class="widget-tcpc widget">
<h3 class="widget-header">Sponsored by</h3>
    <div class="widget-content">
        <a href="http://training.perl.com/" alt="Perl Training" target="_blank"><img src="/i/tcpc.png" width="150" height="50"></a>
    </div>
</div>

<div class="widget-syndication widget">
    <div class="widget-content">
        <ul>
            <li><img src="/mt-static/images/status_icons/feed.gif" alt="Subscribe to feed" width="9" height="9" /> <a href="/pub/atom.xml">Subscribe to this website's feed</a></li>

        </ul>
    </div>
</div>
<div class="widget-archive-monthly widget-archive widget">
    <h3 class="widget-header">Monthly <a href="/pub/archives.html">Archives</a></h3>
    <div class="widget-content">
        <ul>
        
            <li><a href="/pub/2014/02/">February 2014 (1)</a></li>
        
    
        
            <li><a href="/pub/2014/01/">January 2014 (1)</a></li>
        
    
        
            <li><a href="/pub/2013/10/">October 2013 (1)</a></li>
        
    
        
            <li><a href="/pub/2013/01/">January 2013 (1)</a></li>
        
    
        
            <li><a href="/pub/2012/12/">December 2012 (1)</a></li>
        
    
        
            <li><a href="/pub/2012/11/">November 2012 (1)</a></li>
        
    
        
            <li><a href="/pub/2012/10/">October 2012 (2)</a></li>
        
    
        
            <li><a href="/pub/2012/08/">August 2012 (2)</a></li>
        
    
        
            <li><a href="/pub/2012/06/">June 2012 (11)</a></li>
        
    
        
            <li><a href="/pub/2012/05/">May 2012 (18)</a></li>
        
    
        
            <li><a href="/pub/2012/04/">April 2012 (17)</a></li>
        
    
        
            <li><a href="/pub/2012/02/">February 2012 (1)</a></li>
        
    
        
            <li><a href="/pub/2011/12/">December 2011 (1)</a></li>
        
    
        
            <li><a href="/pub/2011/09/">September 2011 (1)</a></li>
        
    
        
            <li><a href="/pub/2011/08/">August 2011 (2)</a></li>
        
    
        
            <li><a href="/pub/2011/06/">June 2011 (1)</a></li>
        
    
        
            <li><a href="/pub/2011/05/">May 2011 (3)</a></li>
        
    
        
            <li><a href="/pub/2011/04/">April 2011 (1)</a></li>
        
    
        
            <li><a href="/pub/2011/03/">March 2011 (1)</a></li>
        
    
        
            <li><a href="/pub/2011/02/">February 2011 (1)</a></li>
        
    
        
            <li><a href="/pub/2011/01/">January 2011 (1)</a></li>
        
    
        
            <li><a href="/pub/2010/11/">November 2010 (1)</a></li>
        
    
        
            <li><a href="/pub/2010/10/">October 2010 (2)</a></li>
        
    
        
            <li><a href="/pub/2010/09/">September 2010 (1)</a></li>
        
    
        
            <li><a href="/pub/2010/08/">August 2010 (3)</a></li>
        
    
        
            <li><a href="/pub/2010/07/">July 2010 (2)</a></li>
        
    
        
            <li><a href="/pub/2010/04/">April 2010 (2)</a></li>
        
    
        
            <li><a href="/pub/2010/03/">March 2010 (4)</a></li>
        
    
        
            <li><a href="/pub/2008/05/">May 2008 (1)</a></li>
        
    
        
            <li><a href="/pub/2008/04/">April 2008 (2)</a></li>
        
    
        
            <li><a href="/pub/2008/03/">March 2008 (1)</a></li>
        
    
        
            <li><a href="/pub/2008/02/">February 2008 (1)</a></li>
        
    
        
            <li><a href="/pub/2008/01/">January 2008 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/12/">December 2007 (2)</a></li>
        
    
        
            <li><a href="/pub/2007/09/">September 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/08/">August 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/07/">July 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/06/">June 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/05/">May 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/04/">April 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/03/">March 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/02/">February 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/01/">January 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/12/">December 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/11/">November 2006 (2)</a></li>
        
    
        
            <li><a href="/pub/2006/10/">October 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/09/">September 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/08/">August 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/07/">July 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/06/">June 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/05/">May 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/04/">April 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/03/">March 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/02/">February 2006 (4)</a></li>
        
    
        
            <li><a href="/pub/2006/01/">January 2006 (4)</a></li>
        
    
        
            <li><a href="/pub/2005/12/">December 2005 (4)</a></li>
        
    
        
            <li><a href="/pub/2005/11/">November 2005 (3)</a></li>
        
    
        
            <li><a href="/pub/2005/10/">October 2005 (2)</a></li>
        
    
        
            <li><a href="/pub/2005/09/">September 2005 (2)</a></li>
        
    
        
            <li><a href="/pub/2005/08/">August 2005 (9)</a></li>
        
    
        
            <li><a href="/pub/2005/07/">July 2005 (8)</a></li>
        
    
        
            <li><a href="/pub/2005/06/">June 2005 (9)</a></li>
        
    
        
            <li><a href="/pub/2005/05/">May 2005 (8)</a></li>
        
    
        
            <li><a href="/pub/2005/04/">April 2005 (7)</a></li>
        
    
        
            <li><a href="/pub/2005/03/">March 2005 (6)</a></li>
        
    
        
            <li><a href="/pub/2005/02/">February 2005 (7)</a></li>
        
    
        
            <li><a href="/pub/2005/01/">January 2005 (6)</a></li>
        
    
        
            <li><a href="/pub/2004/12/">December 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/11/">November 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/10/">October 2004 (5)</a></li>
        
    
        
            <li><a href="/pub/2004/09/">September 2004 (9)</a></li>
        
    
        
            <li><a href="/pub/2004/08/">August 2004 (6)</a></li>
        
    
        
            <li><a href="/pub/2004/07/">July 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/06/">June 2004 (6)</a></li>
        
    
        
            <li><a href="/pub/2004/05/">May 2004 (7)</a></li>
        
    
        
            <li><a href="/pub/2004/04/">April 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/03/">March 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/02/">February 2004 (9)</a></li>
        
    
        
            <li><a href="/pub/2004/01/">January 2004 (7)</a></li>
        
    
        
            <li><a href="/pub/2003/12/">December 2003 (4)</a></li>
        
    
        
            <li><a href="/pub/2003/11/">November 2003 (7)</a></li>
        
    
        
            <li><a href="/pub/2003/10/">October 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2003/09/">September 2003 (6)</a></li>
        
    
        
            <li><a href="/pub/2003/08/">August 2003 (7)</a></li>
        
    
        
            <li><a href="/pub/2003/07/">July 2003 (9)</a></li>
        
    
        
            <li><a href="/pub/2003/06/">June 2003 (9)</a></li>
        
    
        
            <li><a href="/pub/2003/05/">May 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2003/04/">April 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2003/03/">March 2003 (10)</a></li>
        
    
        
            <li><a href="/pub/2003/02/">February 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2003/01/">January 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2002/12/">December 2002 (5)</a></li>
        
    
        
            <li><a href="/pub/2002/11/">November 2002 (9)</a></li>
        
    
        
            <li><a href="/pub/2002/10/">October 2002 (7)</a></li>
        
    
        
            <li><a href="/pub/2002/09/">September 2002 (11)</a></li>
        
    
        
            <li><a href="/pub/2002/08/">August 2002 (8)</a></li>
        
    
        
            <li><a href="/pub/2002/07/">July 2002 (8)</a></li>
        
    
        
            <li><a href="/pub/2002/06/">June 2002 (4)</a></li>
        
    
        
            <li><a href="/pub/2002/05/">May 2002 (6)</a></li>
        
    
        
            <li><a href="/pub/2002/04/">April 2002 (6)</a></li>
        
    
        
            <li><a href="/pub/2002/03/">March 2002 (7)</a></li>
        
    
        
            <li><a href="/pub/2002/02/">February 2002 (5)</a></li>
        
    
        
            <li><a href="/pub/2002/01/">January 2002 (8)</a></li>
        
    
        
            <li><a href="/pub/2001/12/">December 2001 (7)</a></li>
        
    
        
            <li><a href="/pub/2001/11/">November 2001 (5)</a></li>
        
    
        
            <li><a href="/pub/2001/10/">October 2001 (9)</a></li>
        
    
        
            <li><a href="/pub/2001/09/">September 2001 (7)</a></li>
        
    
        
            <li><a href="/pub/2001/08/">August 2001 (13)</a></li>
        
    
        
            <li><a href="/pub/2001/07/">July 2001 (8)</a></li>
        
    
        
            <li><a href="/pub/2001/06/">June 2001 (13)</a></li>
        
    
        
            <li><a href="/pub/2001/05/">May 2001 (11)</a></li>
        
    
        
            <li><a href="/pub/2001/04/">April 2001 (9)</a></li>
        
    
        
            <li><a href="/pub/2001/03/">March 2001 (8)</a></li>
        
    
        
            <li><a href="/pub/2001/02/">February 2001 (8)</a></li>
        
    
        
            <li><a href="/pub/2001/01/">January 2001 (8)</a></li>
        
    
        
            <li><a href="/pub/2000/12/">December 2000 (6)</a></li>
        
    
        
            <li><a href="/pub/2000/11/">November 2000 (10)</a></li>
        
    
        
            <li><a href="/pub/2000/10/">October 2000 (10)</a></li>
        
    
        
            <li><a href="/pub/2000/09/">September 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/2000/08/">August 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/2000/07/">July 2000 (5)</a></li>
        
    
        
            <li><a href="/pub/2000/06/">June 2000 (7)</a></li>
        
    
        
            <li><a href="/pub/2000/05/">May 2000 (7)</a></li>
        
    
        
            <li><a href="/pub/2000/04/">April 2000 (3)</a></li>
        
    
        
            <li><a href="/pub/2000/03/">March 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/2000/02/">February 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/2000/01/">January 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/1999/12/">December 1999 (6)</a></li>
        
    
        
            <li><a href="/pub/1999/11/">November 1999 (6)</a></li>
        
    
        
            <li><a href="/pub/1999/10/">October 1999 (5)</a></li>
        
    
        
            <li><a href="/pub/1999/09/">September 1999 (4)</a></li>
        
    
        
            <li><a href="/pub/1999/08/">August 1999 (3)</a></li>
        
    
        
            <li><a href="/pub/1999/07/">July 1999 (2)</a></li>
        
    
        
            <li><a href="/pub/1999/06/">June 1999 (3)</a></li>
        
    
        
            <li><a href="/pub/1999/04/">April 1999 (1)</a></li>
        
    
        
            <li><a href="/pub/1999/03/">March 1999 (1)</a></li>
        
    
        
            <li><a href="/pub/1999/01/">January 1999 (1)</a></li>
        
    
        
            <li><a href="/pub/1998/12/">December 1998 (1)</a></li>
        
    
        
            <li><a href="/pub/1998/11/">November 1998 (1)</a></li>
        
    
        
            <li><a href="/pub/1998/07/">July 1998 (2)</a></li>
        
    
        
            <li><a href="/pub/1998/06/">June 1998 (1)</a></li>
        
    
        
            <li><a href="/pub/1998/03/">March 1998 (1)</a></li>
        
        </ul>
    </div>
</div>
        
    

<div class="widget-powered widget">
    <div class="widget-content">
        <a href="http://www.movabletype.com/"><img src="/mt-static/images/bug-pbmt-white.png" alt="Powered by Movable Type 5.13-en" width="120" height="75" /></a>
    </div>
</div>



    </div>
</div>






                </div>
            </div>


            <div id="footer">
    <div id="footer-inner">
        <div id="footer-content">
            <div class="widget-powered widget">
                <div class="widget-content">
                    Powered by <a href="http://www.movabletype.com/" rel="generator">Movable Type Pro</a>
                </div>
            </div>

            <div class="widget-creative-commons widget">
                <div class="widget-content">
                    This blog is licensed under a <a href="http://creativecommons.org/licenses/by-nc-nd/3.0/">Creative Commons License</a>.
                </div>
            </div>

        </div>
    </div>
</div>



        </div>
    </div>
</body>
</html>
