<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
    "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" id="sixapart-standard">
<head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="generator" content="Movable Type Pro 5.02" />
<link rel="stylesheet" href="/pub/styles.css" type="text/css" />
<link rel="start" href="/pub/" title="Home" />
<link rel="alternate" type="application/atom+xml" title="Recent Entries" href="/pub/atom.xml" />
<script type="text/javascript" src="/pub/mt.js"></script>

<script type="text/javascript">

 var _gaq = _gaq || [];
 _gaq.push(['_setAccount', 'UA-50555-22']);
 _gaq.push(['_trackPageview']);

 (function() {
   var ga = document.createElement('script'); ga.type =
   'text/javascript'; ga.async = true;
   ga.src = ('https:' == document.location.protocol ? 'https://ssl' :
   'http://www') + '.google-analytics.com/ga.js';
   var s = document.getElementsByTagName('script')[0];
   s.parentNode.insertBefore(ga, s);
 })();

</script>
    
    <link rel="prev bookmark" href="/pub/2003/09/20030907.html" title="This week on Perl 6, week ending 2003-09-07" />
    <link rel="next bookmark" href="/pub/2003/09/perlcookbook.html" title="Cooking with Perl, Part 3" />
    
    
    <title>A Chromosome at a Time with Perl, Part 1 - Perl.com</title>
</head>
<body id="perl-com" class="mt-entry-archive layout-wt">
    <div id="container">
        <div id="container-inner">


            <div id="header">
    <div id="header-inner">
        <div id="header-content">
        <div id="top_advert"> 
<!-- Put any landscape advert in here -->
<a href="http://www.perlfoundation.org/" target="_new">
<img src="/i/tpf_banner.png" width="468" height="60" /></a>
        </div> 



            <div id="header-name"><a href="/pub/" accesskey="1">Perl.com</a></div>
            <div id="header-description"></div>




        </div>
    </div>
</div>



            <div id="content">
                <div id="content-inner">


                    <div id="alpha">
                        <div id="alpha-inner">


                            <div id="entry-1188" class="entry-asset asset hentry">
                                <div class="asset-header">
                                    <h1 id="page-title" class="asset-name entry-title">A Chromosome at a Time with Perl, Part 1</h1>
                                    <div class="asset-meta">
                                        <span class="byline">

                                            By <span class="vcard author">James D. Tisdall</span> on <abbr class="published" title="2003-09-11T00:00:00-08:00">September 11, 2003 12:00 AM</abbr>

                                        </span>


                                    </div>
                                </div>
                                <div class="asset-content entry-content">

                                    <div class="asset-body">
                                        
<!-- sidebar begins -->
<!-- don't move sidebars -->
<!-- sidebar ends -->
<p><i>James D. Tisdall is the author of the soon-to-be-released </i><a href="http://www.oreilly.com/catalog/mperlbio/index.html?CMP=IL7015">Mastering 
    Perl for Bioinformatics</a><i>.</i></p>

<p> For some time now, the use of Perl in biology has been standard practice. 
    Perl remains the most popular language among biologists for a multitude 
    of programming tasks. The same reasons why Perl has been a success story 
    among system administrators, as well as one of the big success stories 
    in the early days of the Web and CGI programming, have also made it 
    the lingua franca of programming in biology, known as <i>bioinformatics</i>. 
</p>
<p> One of the reasons why Perl has been equally well suited to dealing 
    with things like DNA and protein sequence data is that it's so easy 
    to declare and use a string. You just use it, without worrying about 
    allocating memory, or managing memory as the string shrinks or grows. 
    DNA and proteins and other standard biological data are almost always 
    represented in Perl as strings, so this facility with strings translates 
    directly into a facility with DNA and proteins. </p>
<p> For example, say you have a subroutine <code>get_chromosome</code> that returns 
    a string of all the DNA in a human chromosome. In humans, this might 
    be a string about 100Mb in length. This snippet of code calls <code>get_chromosome</code> 
    to initialize a scalar variable, <code>$chromosome1</code>, with the string 
    of DNA sequence data that summarizes human chromosome 1: </p>

<pre><code>$chromosome1 = get_chromosome( 1 );</code></pre>

<p> This programming is as easy as cake. I mean, simple as pie. Well, you 
    know what I mean. </p>
	
	<csperl file="grab" domain="on" record="b/889" template="b/article_sidebar.view">
	
<p> But beneath this wonderful ease of programming lurks a problem. It's 
    a problem that can make your wonderful, intuitive code for tossing around 
    chromosomes and genomes--which looks so elegant in your printout, and 
    which appears so neatly divided into intuitively satisfying, interacting 
    subroutines--an inefficient mess that barely runs at all, when it's 
    not completely clogging up your computer. </p>
<p> So, in this short article I'll show you a handful of tricks 
    that enable you to write code for dealing with large amounts of biological 
    sequence data--in this case, very long strings--while still getting 
    satisfactory speed from the program. </p>
<h3>Memory is the Bottleneck</h3>
<p> What is the problem, exactly? It usually comes down to this: by dealing 
    with very large strings, each one of which uses a significant portion 
    of the main memory that your computer uses to hold a running program, 
    you can easily overtax the amount of main memory available. </p>
<p> When a program on your computer (a <i>process</i> on your Linux, Unix, 
    or Mac OS X computer) runs out of main memory, its performance starts 
    to seriously degrade. It may try to overcome the lack of fast and efficient 
    main memory by enlisting a portion of disk space to hold the part of 
    the running program that it can no longer fit. </p>
<p> But when a program starts writing and reading to and from hard disk 
    memory it can get awfully slow awfully fast. And depending on the nature 
    of the computation, the program may start "thrashing," that is, repeatedly 
    writing and reading large amounts of data between main memory and hard 
    disk. Your elegant program has turned into a greedy, lazy misanthrope 
    that grabs up all the resources available and then seems to just sit 
    there. You've created a monster! </p>
<p> Take the snippet of code above that calls <code>get_chromosome</code>. Without 
    knowing anything more about the subroutine, it's a pretty good bet that 
    it is fetching the 100Mb of data from somewhere, perhaps a disk file, 
    or a relational database, or a web site. To do so, it must be using 
    at least 100Mb of memory. Then, when it returns the data to be stored 
    in <code>$chromosome1</code>, the program uses another 100Mb of memory. Now, 
    perhaps you want to do a regular expression search on the chromosome, 
    saving the desired expression with parentheses that set the special 
    variables <code>$1</code>, <code>$&amp;</code>, and so on. These special 
    variables can be quite large, and that means use of even more memory 
    by your program. </p>
<p> And since this is elegant, simple code you've written, you may well 
    make other copies of the chromosome data or portions of it, in your 
    tenacious hunt for the elusive cure for a deadly disease. The resulting 
    code may be clear, straightforward to understand, and correct--all good 
    and proper things for code to be--but the amount of string copies will 
    land you in the soup. Not only does copying a large string take up memory, 
    but the actual copying can itself be slow, especially if there's a lot 
    of it. </p>
<h3>Space Efficiency</h3>

<!-- sidebar begins -->
<table width="220" border="0" cellspacing="8" cellpadding="4" align="right">
<tr>
<td width="220" valign="top" bgcolor="#efefef">
<div class="secondary">
<h4>More from this author</h4>
<p>
&bull; <a href="/pub/a/2001/11/16/perlbio2.html">Parsing Protein Domains with Perl</a><br />
&bull; <a href="/pub/a/2002/01/02/bioinf.html">Beginning Bioinformatics</a><br />
&bull; <a href="http://www.oreilly.com/news/perlbio_1001.html">Why Biologists Want to Program Computers</a>
</p>
</div>
</td></tr></table>
<!-- sidebar ends -->

<p> You may need to add a new constraint to your program design when you've 
    got a large amount of data in a running program. The constraint is <i>"Use 
    minimal memory."</i> Often, a program that barely runs at all and takes 
    many hours of clogging up the computer, can be rewritten to run in a 
    few minutes by reworking the algorithm so that it uses only a small 
    fraction of the memory. </p>
<p> It's a case of decreasing time by first decreasing space. (Astrophysicists, 
    take note.) </p>
<h3>References</h3>
<p> There's one easy way to cut down on the number of big strings in a program. 
</p>
<p> If you need a subroutine to return a large string, as in the <code>get_chromosome</code> 
    subroutine I've used as an example, you can use <i>references</i> to 
    eliminate some of this memory usage. </p>
<p> The practice of passing references to a subroutine is familiar to experienced 
    Perl programmers. In our example, we can rewrite the subroutine so that 
    the return value is placed into a string that is passed in as an argument. 
    But we don't pass a copy of the string--we pass a reference to the string, 
    which takes almost no additional space, and which still enables the 
    subroutine to provide the entire chromosome 1 DNA to the calling program. 
    Here's an example: </p>

<pre><code>load_chromosome( 1, \$chromosome1 );</code></pre>

<p> This new subroutine has two arguments. The <code>1</code> presumably will tell the 
    subroutine which human chromosome we want (we want the biggest human 
    chromosome, chromosome 1). </p>
<p> The second argument is a reference to a scalar variable. Inside the 
    subroutine, the reference is most likely used to initialize an argument 
    like this: </p>

<pre><code>my($chromnumber, $chromref) = @_;</code></pre>

<p> And then the DNA data is put into the string by calling it <code>$$chromref</code>, 
    for instance like so: </p>
	
<pre><code>$$chromref = 'ACGTGTGAACGGA';</code></pre>

<p> No return value is needed. After the subroutine call, the main program 
    will find that the contents of <code>$chromosome1</code> have changed, and now consist 
    of "ACGTGTGAACGGA." (Of course, a chromosome is much longer than this 
    little fragment.) </p>
<p> Using references is also a great way to pass a large amount of data 
    <i>into</i> a subroutine without making copies of it. In this case, 
    however, the fact that the subroutine can change the contents of the 
    referenced data is something to watch out for. Sometimes you just want 
    a subroutine to get to use the data, but you expect the variable containing 
    the data to still have the same data after the subroutine gets a look 
    at it. So you have to watch what you do when you're passing references 
    to data into a subroutine, and make sure you know what you want. </p>
<h3>Managing Memory with Buffers</h3>
<p> One of the most efficient ways to deal with very large strings is to 
    deal with them a little at a time. </p>
<p> Here's an example of a program for searching an entire chromosome for 
    a particular 12-base pattern, using very little memory. (A <i>base</i> 
    is one of the four molecules that are the principal building blocks 
    of DNA. The four bases are represented in Perl strings as the characters 
    A, C, G, and T. You'll often hear biologists talking about "megabases" 
    instead of "megabytes" in a string. If you hear that, you're probably 
    talking to a bioinformatician.) </p>
<p> When writing a program that will search for any regular expression in 
    a chromosome, it's hard to see how you could avoid putting the whole 
    chromosome in a string. But very often there's a limit to the size of 
    what you're searching for. In this program, I'm looking for the 12-base 
    pattern "ACGTACGTACGT." And I'm going to get the chromosome data from 
    a disk file. </p>
<p> My trick is going to be to just read in the chromosome data a line or 
    two at a time, search for the pattern, and then <i>reuse</i> the memory 
    to read in the next line or two of data. </p>
<p> The extra work I have to do in programming is, first, I need to 
    keep track myself of how much of the data has been read in, so I can 
    report the locations in the chromosome of successful searches. Second, 
    I need to keep aware that my pattern might start at the end of one line 
    and complete at the beginning of the next line, so I need to make sure 
    I search across line breaks as well as within lines of data from the 
    input file. </p>


<p> Here's a small program that reads in a FASTA file given as an argument 
    on the command line and searches for my pattern in any amount of DNA--a 
    whole chromosome, a whole genome, even all known genetic data, just 
    assuming that the data is in a FASTA file named in the command line. 
    I'll call my program <code>find_fragment</code>, and assuming the DNA is in 
    a FASTA file called <code>human.dna</code>, I'll call it like so: </p>

<pre><code>[tisdall@coltrane]$ perl find_fragment human.dna</code></pre>

<p> For testing purposes I made a very short FASTA DNA file, <code>human.dna</code>, 
    which contains: </p>
<pre><code>&gt; human dna--Not!  The fragment ACGTACGTACGT appears at positions 10, 40, and 98
AAAAAAAAAACGTACGTACGTCCGCGCGCGCGCGCGCGCACGTACGTACG
TGGGGGGGGGGGGGGGCCCCCCCCCCGGGGGGGGGGGGAAAAAAAAAACG
TACGTACGTTTTTTTTTTTTTTTTTTTTTTTTTTT</code></pre>
<p> Here's the code for the program <code>find_fragment</code>: </p>
<pre><code>#!/usr/bin/perl

#
# find_fragment : find 'ACGTACGTACGT' in a very large DNA FASTA file 
# using minimal memory
#
#  N.B. This example program does no checking of the input to ensure 
#       that it is DNA data in FASTA format; it just assumes that 
#       it is. This program also assumes there is just one FASTA
#       record in the input file.
#
#  Copyright (c) 2003 James Tisdall
#

use warnings;
use strict;
use Carp;

# Make sure the program is called with one argument, presumably a 
# FASTA file of DNA
my $USAGE = "perl find_fragment file.FASTA";
unless(@ARGV == 1) { croak "$USAGE:$!\n" }

# $fragment: the pattern to search for
# $fraglen:  the length of $fragment
# $buffer:   a buffer to hold the DNA from the input file
# $position: the position of the buffer in the total DNA
my($fragment, $fraglen, $buffer, $position) = ('ACGTACGTACGT', 12, '', 0);

# The first line of a FASTA file is a header and begins with '&gt;'
my $header = &lt;&gt;;

# Get the first line of DNA data, to start the ball rolling
$buffer = &lt;&gt;;
chomp $buffer;

# The remaining lines are DNA data ending with newlines
while(my $newline = &lt;&gt;) {

    # Add the new line to the buffer
    chomp $newline;
    $buffer .= $newline;

    # Search for the DNA fragment, which has a length of 12
    # (Report the character at string position 0 as being at position 1, 
    # as usual in biology)
    while($buffer =~ /$fragment/gi) {
        print "Found $fragment at position ", $position + $-[0] + 1, "\n";
    }

    # Reset the position counter (will be true after you reset the buffer, next)
    $position = $position + length($buffer) - $fraglen + 1;

    # Discard the data in the buffer, except for a portion at the end
    # so patterns that appear across line breaks are not missed
    $buffer = substr($buffer, length($buffer) - $fraglen + 1, $fraglen - 1);
}</code></pre>

<p> Here's the output of running the command <code>perl find_fragment human.dna</code>: 
</p>
<pre><code>Found ACGTACGTACGT at position 10
Found ACGTACGTACGT at position 40
Found ACGTACGTACGT at position 98</code></pre>

<h3>How the Code Works</h3>
<p> After the highly recommended <code>use strict</code> and <code>use warnings</code> are turned on, and 
    the Carp module is loaded so the program can "croak" when needed, the 
    program variables are declared and initialized. </p>
<p> The first line of the FASTA file is a header and is not needed here, 
    so it's read and not used. Then the first line of DNA data is read into 
    the buffer and its <code>newline</code> character is removed. I start with this because 
    I want to search for the fragment even if it is broken by new lines, 
    so I'll have to look at least at the first two lines; here I get the 
    first line, and in the while loop that follows I'll start by adding 
    the second line to the buffer. </p>
<p> Then the while loop, which does the main work of the program, starts 
    reading in the next line of the FASTA file named on the command line, 
    in this case the FASTA file <code>human.dna</code>. The <code>newline</code> is removed with 
    "chomp," and the new line is added to the buffer. </p>
<p> Then comes the short while loop that does the regular expression pattern 
    match of the <code>$fragment</code> in the <code>$buffer</code>. It has modifiers "g" for <i>g</i>lobal 
    search (the fragment may appear more than once in the buffer); and "i" 
    for case <i>i</i>nsensitive search, that is, either uppercase or lowercase DNA data (e.g. ACGT or acgt). </p>
<p> When the fragment is found the program simply prints out the position. 
    <code>$position</code> holds the position of the beginning of the buffer in the total 
    DNA, and is something I have to keep track of. <code>$-[0]</code> is a special variable 
    that gives the offset of the last successful pattern match in the string. 
    I also add 1, because biologists always say that the first base in a 
    sequence of DNA is at position 1, whereas Perl says that the first character 
    in a string is at position 0. So I add 1 to the Perl position to get 
    the biologist's position. </p>
<p> The last two lines of code reset the buffer by eliminating the beginning 
    part of it, and then adjust the position counter accordingly. The buffer 
    is shortened so that it just keeps the part at the very end that might 
    be part of a pattern match that crosses over the lines of the input 
    file. This would be the tail part of the buffer that is just one base 
    shorter than the length of the fragment. </p>
<p> In this way, the program keeps at most two lines' worth of DNA in <code>$buffer</code>, 
    but still manages to search the entire genome (or chromosome or whatever 
    is in the FASTA file) for the fragment. It performs very quickly, compared 
    to a program that reads in a whole genome and blows out the memory in 
    the process. </p>
<h3>When You Should Bother</h3>
<p> A space-inefficient program might well work fine on your better computers, 
    but it won't work well at all when you need to run it on another computer 
    with less main memory installed. Or, it might work fine on the fly genome, 
    but slow to a crawl on the human genome. </p>
<p> The rule of thumb is that if you know you'll be dealing with large 
    data sets, consider the amount of space your program uses as an 
    important constraint when designing and coding. Then you won't have 
    to go back and redo the entire program when a large amount of DNA gets 
    thrown at you. </p>
<p><i>Editor's note: Stay tuned for part two in this two-part series later 
    this month. In it, James will take a more in-depth look at space efficiency, 
    and include a more general version of a program that uses a buffer. 
    In particular, part two will cover running subroutines with minimal space, eliminating subroutines altogether, and sequence motifs with 
    bounded lengths.</i></p>

<hr size="1" noshade="noshade" />

<p> O'Reilly &amp; Associates will soon release (September 2003) <a href="http://www.oreilly.com/catalog/mperlbio/index.html?CMP=IL7015">Mastering 
    Perl for Bioinformatics</a>. </p>
<ul>
    <li>
        <p> <a href="http://www.oreilly.com/catalog/mperlbio/chapter/index.html?CMP=IL7015">Sample Chapter 9, Introduction to Bioperl</a>, is 
            available free online. </p>
    </li>
    <li>
        <p> You can also look at the <a href="http://www.oreilly.com/catalog/mperlbio/toc.html?CMP=IL7015">Table of Contents</a>, the 
            <a href="http://www.oreilly.com/catalog/mperlbio/inx.html?CMP=IL7015">Index</a>, and the <a href="http://www.oreilly.com/catalog/mperlbio/desc.html?CMP=IL7015">Full Description</a> 
            of the book. </p>
    </li>
    <li>
        <p> For more information, or to order the book, <a href="http://www.oreilly.com/catalog/mperlbio/index.html?CMP=IL7015">click 
            here</a>. </p>
    </li>
</ul>
                                    </div>


                                </div>
                                <div class="asset-footer">

    
                                    <div class="entry-categories">
                                        <h4>Categories<span class="delimiter">:</span></h4>
                                        <ul>
                                            <li><a href="/pub/biology/" rel="tag">Biology</a></li>
                                        </ul>
                                    </div>
    


                                    <div class="entry-tags">
                                        <h4>Tags<span class="delimiter">:</span></h4>
                                        <ul>
                                            <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?blog_id=2&amp;tag=bioinformatics&amp;limit=20';return false;" rel="tag">bioinformatics</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?blog_id=2&amp;tag=biological%20sequence%20data&amp;limit=20';return false;" rel="tag">biological sequence data</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?blog_id=2&amp;tag=James%20Tisdall&amp;limit=20';return false;" rel="tag">James Tisdall</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?blog_id=2&amp;tag=perl%20for%20bioinformatics&amp;limit=20';return false;" rel="tag">perl for bioinformatics</a><span class="delimiter">,</span></li> <li><a href="javascript:void(0)" onclick="location.href='http://www.perl.com/mt-search.cgi?blog_id=2&amp;tag=Tisdall&amp;limit=20';return false;" rel="tag">Tisdall</a></li>
                                        </ul>
                                    </div>

                                </div>
                            </div>


                    
                    


                        </div>
                    </div>


                    <div id="beta">
    <div id="beta-inner">


    
    <div class="widget-what-is-perl widget">
    <div class="widget-content widget-content-what-is-perl">
       Visit the home of the  Perl programming language: <a href="http://www.perl.org/">Perl.org</a
    </div>
</div>
<div class="widget-find-out-more widget-archives widget">
    <div class="widget-content">
        <ul>
            <li><a href="http://www.perl.org/get.html">Download</a></li>
            <li><a href="http://perldoc.perl.org/">Documentation</a></li>
            <li><a href="http://blogs.perl.org/">Perl Bloggers</a></li>
            <li><a href="http://news.perlfoundation.org/">Foundation News</a></li>
        </ul>
    </div>
</div><div class="widget-tcpc widget">
<h3 class="widget-header">Sponsored by</h3>
    <div class="widget-content">
        <a href="http://training.perl.com/" alt="Perl Training" target="_blank"><img src="/i/tcpc.png" width="150" height="50"></a>
    </div>
</div>

<div class="widget-syndication widget">
    <div class="widget-content">
        <ul>
            <li><img src="/mt-static/images/status_icons/feed.gif" alt="Subscribe to feed" width="9" height="9" /> <a href="/pub/atom.xml">Subscribe to this website's feed</a></li>

        </ul>
    </div>
</div>
<div class="widget-archive-monthly widget-archive widget">
    <h3 class="widget-header">Monthly <a href="/pub/archives.html">Archives</a></h3>
    <div class="widget-content">
        <ul>
        
            <li><a href="/pub/2010/07/">July 2010 (1)</a></li>
        
    
        
            <li><a href="/pub/2008/05/">May 2008 (1)</a></li>
        
    
        
            <li><a href="/pub/2008/04/">April 2008 (2)</a></li>
        
    
        
            <li><a href="/pub/2008/03/">March 2008 (1)</a></li>
        
    
        
            <li><a href="/pub/2008/02/">February 2008 (1)</a></li>
        
    
        
            <li><a href="/pub/2008/01/">January 2008 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/12/">December 2007 (2)</a></li>
        
    
        
            <li><a href="/pub/2007/09/">September 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/08/">August 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/07/">July 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/06/">June 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/05/">May 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/04/">April 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/03/">March 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/02/">February 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2007/01/">January 2007 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/12/">December 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/11/">November 2006 (2)</a></li>
        
    
        
            <li><a href="/pub/2006/10/">October 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/09/">September 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/08/">August 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/07/">July 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/06/">June 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/05/">May 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/04/">April 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/03/">March 2006 (1)</a></li>
        
    
        
            <li><a href="/pub/2006/02/">February 2006 (4)</a></li>
        
    
        
            <li><a href="/pub/2006/01/">January 2006 (4)</a></li>
        
    
        
            <li><a href="/pub/2005/12/">December 2005 (4)</a></li>
        
    
        
            <li><a href="/pub/2005/11/">November 2005 (3)</a></li>
        
    
        
            <li><a href="/pub/2005/10/">October 2005 (2)</a></li>
        
    
        
            <li><a href="/pub/2005/09/">September 2005 (2)</a></li>
        
    
        
            <li><a href="/pub/2005/08/">August 2005 (9)</a></li>
        
    
        
            <li><a href="/pub/2005/07/">July 2005 (8)</a></li>
        
    
        
            <li><a href="/pub/2005/06/">June 2005 (9)</a></li>
        
    
        
            <li><a href="/pub/2005/05/">May 2005 (8)</a></li>
        
    
        
            <li><a href="/pub/2005/04/">April 2005 (7)</a></li>
        
    
        
            <li><a href="/pub/2005/03/">March 2005 (6)</a></li>
        
    
        
            <li><a href="/pub/2005/02/">February 2005 (7)</a></li>
        
    
        
            <li><a href="/pub/2005/01/">January 2005 (6)</a></li>
        
    
        
            <li><a href="/pub/2004/12/">December 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/11/">November 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/10/">October 2004 (5)</a></li>
        
    
        
            <li><a href="/pub/2004/09/">September 2004 (9)</a></li>
        
    
        
            <li><a href="/pub/2004/08/">August 2004 (6)</a></li>
        
    
        
            <li><a href="/pub/2004/07/">July 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/06/">June 2004 (6)</a></li>
        
    
        
            <li><a href="/pub/2004/05/">May 2004 (7)</a></li>
        
    
        
            <li><a href="/pub/2004/04/">April 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/03/">March 2004 (8)</a></li>
        
    
        
            <li><a href="/pub/2004/02/">February 2004 (9)</a></li>
        
    
        
            <li><a href="/pub/2004/01/">January 2004 (7)</a></li>
        
    
        
            <li><a href="/pub/2003/12/">December 2003 (4)</a></li>
        
    
        
            <li><a href="/pub/2003/11/">November 2003 (7)</a></li>
        
    
        
            <li><a href="/pub/2003/10/">October 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2003/09/">September 2003 (6)</a></li>
        
    
        
            <li><a href="/pub/2003/08/">August 2003 (7)</a></li>
        
    
        
            <li><a href="/pub/2003/07/">July 2003 (9)</a></li>
        
    
        
            <li><a href="/pub/2003/06/">June 2003 (9)</a></li>
        
    
        
            <li><a href="/pub/2003/05/">May 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2003/04/">April 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2003/03/">March 2003 (10)</a></li>
        
    
        
            <li><a href="/pub/2003/02/">February 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2003/01/">January 2003 (8)</a></li>
        
    
        
            <li><a href="/pub/2002/12/">December 2002 (5)</a></li>
        
    
        
            <li><a href="/pub/2002/11/">November 2002 (9)</a></li>
        
    
        
            <li><a href="/pub/2002/10/">October 2002 (7)</a></li>
        
    
        
            <li><a href="/pub/2002/09/">September 2002 (11)</a></li>
        
    
        
            <li><a href="/pub/2002/08/">August 2002 (8)</a></li>
        
    
        
            <li><a href="/pub/2002/07/">July 2002 (8)</a></li>
        
    
        
            <li><a href="/pub/2002/06/">June 2002 (4)</a></li>
        
    
        
            <li><a href="/pub/2002/05/">May 2002 (6)</a></li>
        
    
        
            <li><a href="/pub/2002/04/">April 2002 (6)</a></li>
        
    
        
            <li><a href="/pub/2002/03/">March 2002 (7)</a></li>
        
    
        
            <li><a href="/pub/2002/02/">February 2002 (5)</a></li>
        
    
        
            <li><a href="/pub/2002/01/">January 2002 (8)</a></li>
        
    
        
            <li><a href="/pub/2001/12/">December 2001 (7)</a></li>
        
    
        
            <li><a href="/pub/2001/11/">November 2001 (5)</a></li>
        
    
        
            <li><a href="/pub/2001/10/">October 2001 (9)</a></li>
        
    
        
            <li><a href="/pub/2001/09/">September 2001 (7)</a></li>
        
    
        
            <li><a href="/pub/2001/08/">August 2001 (13)</a></li>
        
    
        
            <li><a href="/pub/2001/07/">July 2001 (8)</a></li>
        
    
        
            <li><a href="/pub/2001/06/">June 2001 (13)</a></li>
        
    
        
            <li><a href="/pub/2001/05/">May 2001 (11)</a></li>
        
    
        
            <li><a href="/pub/2001/04/">April 2001 (9)</a></li>
        
    
        
            <li><a href="/pub/2001/03/">March 2001 (8)</a></li>
        
    
        
            <li><a href="/pub/2001/02/">February 2001 (8)</a></li>
        
    
        
            <li><a href="/pub/2001/01/">January 2001 (8)</a></li>
        
    
        
            <li><a href="/pub/2000/12/">December 2000 (6)</a></li>
        
    
        
            <li><a href="/pub/2000/11/">November 2000 (10)</a></li>
        
    
        
            <li><a href="/pub/2000/10/">October 2000 (10)</a></li>
        
    
        
            <li><a href="/pub/2000/09/">September 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/2000/08/">August 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/2000/07/">July 2000 (5)</a></li>
        
    
        
            <li><a href="/pub/2000/06/">June 2000 (7)</a></li>
        
    
        
            <li><a href="/pub/2000/05/">May 2000 (7)</a></li>
        
    
        
            <li><a href="/pub/2000/04/">April 2000 (3)</a></li>
        
    
        
            <li><a href="/pub/2000/03/">March 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/2000/02/">February 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/2000/01/">January 2000 (2)</a></li>
        
    
        
            <li><a href="/pub/1999/12/">December 1999 (6)</a></li>
        
    
        
            <li><a href="/pub/1999/11/">November 1999 (6)</a></li>
        
    
        
            <li><a href="/pub/1999/10/">October 1999 (5)</a></li>
        
    
        
            <li><a href="/pub/1999/09/">September 1999 (4)</a></li>
        
    
        
            <li><a href="/pub/1999/08/">August 1999 (3)</a></li>
        
    
        
            <li><a href="/pub/1999/07/">July 1999 (2)</a></li>
        
    
        
            <li><a href="/pub/1999/06/">June 1999 (3)</a></li>
        
    
        
            <li><a href="/pub/1999/04/">April 1999 (1)</a></li>
        
    
        
            <li><a href="/pub/1999/03/">March 1999 (1)</a></li>
        
    
        
            <li><a href="/pub/1999/01/">January 1999 (1)</a></li>
        
    
        
            <li><a href="/pub/1998/12/">December 1998 (1)</a></li>
        
    
        
            <li><a href="/pub/1998/11/">November 1998 (1)</a></li>
        
    
        
            <li><a href="/pub/1998/07/">July 1998 (2)</a></li>
        
    
        
            <li><a href="/pub/1998/06/">June 1998 (1)</a></li>
        
    
        
            <li><a href="/pub/1998/03/">March 1998 (1)</a></li>
        
        </ul>
    </div>
</div>
        
    

<div class="widget-powered widget">
    <div class="widget-content">
        <a href="http://www.movabletype.com/"><img src="/mt-static/images/bug-pbmt-white.png" alt="Powered by Movable Type 5.02" width="120" height="75" /></a>
    </div>
</div>



    </div>
</div>






                </div>
            </div>


            <div id="footer">
    <div id="footer-inner">
        <div id="footer-content">
            <div class="widget-powered widget">
                <div class="widget-content">
                    Powered by <a href="http://www.movabletype.com/" rel="generator">Movable Type Pro</a>
                </div>
            </div>

        </div>
    </div>
</div>



        </div>
    </div>
</body>
</html>
